go/vt/vtgate/engine/recurse_cte.go

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2024 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package engine
+
+import (
+	"context"
+
+	"vitess.io/vitess/go/sqltypes"
+	querypb "vitess.io/vitess/go/vt/proto/query"
+)
+
+// RecurseCTE is used to represent recursive CTEs
+// Seed is used to represent the non-recursive part that initializes the result set.
+// It's result are then used to start the recursion on the Term side
+// The values being sent to the Term side are stored in the Vars map -
+// the key is the bindvar name and the value is the index of the column in the recursive result
+type RecurseCTE struct {
+	Seed, Term Primitive
+
+	Vars map[string]int
+}
+
+var _ Primitive = (*RecurseCTE)(nil)
+
+func (r *RecurseCTE) TryExecute(ctx context.Context, vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool) (*sqltypes.Result, error) {
+	res, err := vcursor.ExecutePrimitive(ctx, r.Seed, bindVars, wantfields)
+	if err != nil {
+		return nil, err
+	}
+
+	// recurseRows contains the rows used in the next recursion
+	recurseRows := res.Rows
+	joinVars := make(map[string]*querypb.BindVariable)
+	for len(recurseRows) > 0 {
+		// copy over the results from the previous recursion
+		theseRows := recurseRows
+		recurseRows = nil
+		for _, row := range theseRows {
+			for k, col := range r.Vars {
+				joinVars[k] = sqltypes.ValueBindVariable(row[col])
+			}
+			rresult, err := vcursor.ExecutePrimitive(ctx, r.Term, combineVars(bindVars, joinVars), false)
+			if err != nil {
+				return nil, err
+			}
+			recurseRows = append(recurseRows, rresult.Rows...)
+			res.Rows = append(res.Rows, rresult.Rows...)
+		}
+	}
+	return res, nil
```

Comment:
coercion is not done, either we add the support or fail on type mismatch,
this only needs to be done with result of seed and first result of term

Reply:
This could lead to wrong output. I am suggesting we at least add a failing condition check on field types.